sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/VectorReservePolicy.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.vectorized;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import org.apache.spark.sql.internal.SQLConf;
+import org.apache.spark.unsafe.array.ByteArrayMethods;
+
+public abstract class VectorReservePolicy {
+
+  protected int defaultCapacity;
+
+  /**
+   * Upper limit for the maximum capacity for this column.
+   */
+  @VisibleForTesting
+  protected int MAX_CAPACITY = ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH;
+
+  /**
+   * True if this column has default values. Return the default values instead of NULL when the
+   * corresponding columns are not present in storage. We can not reset the data of column vectors
+   * that has default values.
+   */
+  protected boolean hasDefaultValue = false;
+
+  abstract int nextCapacity(int requiredCapacity);
+
+  boolean shouldCleanData() {
+    return false;
+  }
+}
+
+class DefaultVectorReservePolicy extends VectorReservePolicy {
+
+  private int hugeThreshold;
+  private double hugeReserveRatio;
+  private int currentCapacity;
+
+  DefaultVectorReservePolicy(int defaultCapacity) {
+    this.defaultCapacity = defaultCapacity;
+    this.currentCapacity = defaultCapacity;
+    SQLConf conf = SQLConf.get();
+    hugeThreshold = conf.vectorizedHugeVectorThreshold();
+    hugeReserveRatio = conf.vectorizedHugeVectorReserveRatio();
+  }
+
+  @Override
+  public int nextCapacity(int requiredCapacity) {
+    if (hugeThreshold < 0 || requiredCapacity < hugeThreshold) {
+      currentCapacity = (int) Math.min(MAX_CAPACITY, requiredCapacity * 2L);
+    } else {
+      currentCapacity = (int) Math.min(MAX_CAPACITY, requiredCapacity * hugeReserveRatio);
```

Comment:
This may lead to conter-intuitive behavior for example If the threshold is 1000, then require 999 might look more memory-hungry than 1001. How about we make it 
```
(requiredCapacity - hugeThreshold) * hugeReserveRatio + hugeThreshold * 2L
``` 

to be more consistent

Reply:
Thanks @liuzqt  for your review.

For  `(requiredCapacity - hugeThreshold) * hugeReserveRatio + hugeThreshold * 2L` is equal to `requiredCapacity * hugeReserveRatio + hugeThreshold * (2L - hugeReserveRatio)`. If hugeThreshold= 100M, and the requiredCapacity = 1000, before this change, spark will only allocate 2000 bytes while after this change, spark will allocate 1000 * 1.2 + 100M * 0.8 ~= 80M bytes, will small column vectors take up too much memory?

Reply:
Oh sorry for the confusion, the above formula only apply for the `else` case (i.e, `requiredCapacity >= hugeThreshold`), for `requiredCapacity < hugeThreshold` it's still `requiredCapacity * 2L` as in your code.

```
if (hugeThreshold < 0 || requiredCapacity < hugeThreshold) {
    currentCapacity = (int) Math.min(MAX_CAPACITY, requiredCapacity * 2L);
} else {
    // we only change this
    currentCapacity = (int) Math.min(MAX_CAPACITY, (requiredCapacity - hugeThreshold) * hugeReserveRatio + hugeThreshold * 2L);
}
```

the idea is that for `requiredCapacity >= hugeThreshold`, the exceeding part we multiply by `hugeReserveRatio`

Reply:
Thanks for your idea, I have updated the formula.